### ISSUE TITLE
"Task was cancelled" exception when testing CoProcessFunction

### ISSUE THREAD

**Issue by whipit** [ottogroup/flink-spector#71]
Does flink spector work for testing CoProcessFunction operators?  When attempting to test it, I keep getting a 

`java.util.concurrentCancellationException:  Task was cancelled.`

Notably, the exception seems to occur because in io.flinkspector.core.runtime.OutputSubscriber class in [method getNextMessage()](https://github.com/ottogroup/flink-spector/blob/5479ad91979e252752dcae0fe0089c0c9f8bdc11/flinkspector-core/src/main/java/io/flinkspector/core/runtime/OutputSubscriber.java#L74).    The method appears to throw an InterruptedException, with then causes it to return an empty byte array, which then causes the job to end.

```
    public byte[] getNextMessage() {
        try {
            return queue.poll(TIMEOUT, TimeUnit.MINUTES);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            //Thread interrupted stop waiting and signal it to the subscriber
            return new byte[0];
        }
    }
```

This then causes the [OutputHandler to finish](https://github.com/ottogroup/flink-spector/blob/5479ad91979e252752dcae0fe0089c0c9f8bdc11/flinkspector-core/src/main/java/io/flinkspector/core/runtime/OutputHandler.java#L158)...

```
        if (bytes.length == 0) {
            //the subscriber has been cancelled from outside quietly finish the process
            return Action.FINISH;
        }
```

My CoProcessFunction is very (intentionally) simple:

```
public class DummyJoinFunction extends CoProcessFunction<MyType, MyOtherType, MyType> {
     
     public void processElement1(MyType value, Context ctx, Collector<MyType> out)  {
          // just pass the input to the collector;
          out.collect(value);
     }

     public void processElement1(MyOtherType value, Context ctx, Collector<MyType> out)  {
          // do nothing here;
     }
}
```

And my test class:

```
public class TestDummyJoinFunction extends DataStreamTestBase {
      
       @Test
       public void testTheFunction() {
             Input<MyType> myTypeInput = InputBuilder.startsWith(new MyType(1)).emit(new MyType(2));
             KeyedStream<MyType, Integer> myInputStream = this.createTestStream(myTypeInput).keyBy(new MyKeySelector());

             Input<MyOtherType> myOtherTypeInput = InputBuilder.startsWith(new MyOtherType(1)).emit(new MyOtherType(2));
             KeyedStream<MyOtherType, Integer> myOtherInputStream = this.createTestStream(myOtherTypeInput).keyBy(new MyOtherKeySelector());

             DataStream outputStream = myInputStream.connect(myOtherInputStream).process (new DummyJoinFunction());

             assertStream (outputStream, new MyOutputMatcher());             
       }
}
```

**Comment by lofifnc** [comment]
I think I have an idea what's the problem. I will debug this during the weekend. Thanks for pointing this out.

**Comment by whipit** [comment]
I got it working - but by changing my design a bit.   I missed mentioning a key design in my Test classes that I think is causing flink-spector to not work.   Essentially, I am using nested non-static classes to implement a Given-When-Then (GWT) testing pattern.   So I have something like this:

```
public class GivenMyInputStream extends DataStreamTestBase {
       KeyedStream<MyInput, Integer> myInputStream;
       @Before
       public void setup() {
             Input<MyType> myTypeInput = InputBuilder.startsWith(new MyType(1)).emit(new MyType(2));
             myInputStream = this.createTestStream(myTypeInput).keyBy(new MyKeySelector());           
       }

       public class WhenJoinedToMyOtherInputStream extends DataStreamTestBase {
             KeyedStream<MyInput, Integer> myOtherInputStream;

             @Before
             public void setup() {
                       Input<MyOtherType> myOtherTypeInput = InputBuilder.startsWith(new MyOtherType(1)).emit(new MyOtherType(2));
                      myOtherInputStream = this.createTestStream(myOtherTypeInput).keyBy(new MyOtherKeySelector());
             }

              @Test
              public void ensureJoinWorks() {                             
                            DataStream outputStream = myInputStream.connect(myOtherInputStream).process (new DummyJoinFunction());

                              assertStream (outputStream, new MyOutputMatcher());             
              }
      }
}
```

This pattern does not play well with flink spector -- likely because my 2 streams are in 2 different classes (one nested) both extending from DataStreamTestBase.

I would still like to understand why exactly, but I am inclined to not call this an "Issue" but rather "User error" if its hard to fix.

**Comment by lofifnc** [comment]
The DataStreamTestBase is just a template for setting up JUnit test suites. It takes care of setting up and executing the test. Each instance will set up it's own [DataStreamTestEnvironment](https://github.com/ottogroup/flink-spector/blob/master/flinkspector-datastream/src/main/java/io/flinkspector/datastream/DataStreamTestEnvironment.java) in the background. I tried to make specifing test as concise as possible so createTestStream() will register a source with the Environment in the Background. So you're working with two distinct test environments. However the `Input` object is portable and reusable. The stream is registered with one environment.

 If you want to combine DataStreamTestBases you can take a look at the (DataStreamTestBase)[https://github.com/ottogroup/flink-spector/blob/master/flinkspector-datastream/src/main/java/io/flinkspector/datastream/DataStreamTestBase.java], and how the DataStreamTestEnvironment is utilized. You can find how a test is setup and executed there. The `testEnv` variable in the `DataStreamTestBase` is also accessible so you can make sure both are using the same environment.

**Comment by whipit** [comment]
Thank you!  That makes sense.   I'll close this issue.

**Comment by whipit** [comment]
Related (but different) question:   Does the framework support running multiple tests (different implementations and instances of DataStreamTestBase) in parallel, or even in the same JUnit JVM?

I am trying to do this, but am getting flickering results.   For instance, I have 2 nearly identical implementations of DataStreamTestBase.  For simplicity sake, both are testing exactly the same input data.

When I run the tests, one of them succeeds whereas the other one fails.  The latter fails because the output stream seems to have incorrect data -- specifically it seems to have data from the first instance.

However, if I run the test in debug mode, and pause at a break point long enough, then both tests succeed.

**Comment by whipit** [comment]
to keep it clean, I opened a separate issue for the shared state problem.  

https://github.com/ottogroup/flink-spector/issues/72
